Source/WebKit/UIProcess/Cocoa/UIDelegate.mm

```diff
@@ -2269,4 +2272,28 @@ static _WKXRSessionFeatureFlags toWKXRSessionFeatureFlags(const PlatformXR::Devi
 
 #endif // ENABLE(WEBXR)
 
+void UIDelegate::UIClient::geometryDidChange(WebPageProxy&, WebCore::FloatSize containerSize, WebCore::FloatBoxExtent contentInsets, WebCore::FloatPoint contentOffset, WebCore::FloatSize contentSize)
+{
+    RefPtr uiDelegate = m_uiDelegate.get();
+    if (!uiDelegate || !uiDelegate->m_delegateMethods.webViewGeometryDidChange)
+        return;
+
+    auto delegate = (id<WKUIDelegateInternal>)uiDelegate->m_delegate.get();
+    if (!delegate)
+        return;
+
+#if PLATFORM(IOS_FAMILY)
+    UIEdgeInsets insets = UIEdgeInsetsMake(contentInsets.top(), contentInsets.left(), contentInsets.bottom(), contentInsets.right());
+#else
+    NSEdgeInsets insets = NSEdgeInsetsMake(contentInsets.top(), contentInsets.left(), contentInsets.bottom(), contentInsets.right());
+#endif
+
+    RetainPtr geometry = adoptNS([[WKScrollGeometry alloc] initWithContainerSize:containerSize contentInsets:insets contentOffset:contentOffset contentSize:contentSize]);
+
+    if (![delegate respondsToSelector:@selector(_webView:geometryDidChange:)])
+        return;
+
+    [delegate _webView:uiDelegate->m_webView.get().get() geometryDidChange:geometry.get()];
```

Comment:
I know you check for equality later on, in the Swift code, but it seems wrong to call the delegate at all if the geometry is unchanged. Can we move the equality check up here to avoid unnecessarily calling the delegate?

Reply:
(or even before here, in WebPageProxy itself)

Reply:
> I know you check for equality later on, in the Swift code, but it seems wrong to call the delegate at all if the geometry is unchanged. Can we move the equality check up here to avoid unnecessarily calling the delegate?

yup, will fix. (I do have to leave the Swift code still though since it does a stricter equality check of the transformed value).